File: node/solana-cli-tool.js
#!/usr/bin/env node

const { Keypair, Connection, LAMPORTS_PER_SOL, PublicKey, SystemProgram, Transaction } = require('@solana/web3.js');
const fs = require('fs');
const path = require('path');
const os = require('os');
const argv = require('yargs/yargs')(process.argv.slice(2)).argv;

const CONFIG_DIR = path.join(os.homedir(), '.config', 'solana');
const KEYPAIR_PATH = path.join(CONFIG_DIR, 'id.json');

// Ensure the config directory exists
if (!fs.existsSync(CONFIG_DIR)) {
    fs.mkdirSync(CONFIG_DIR, { recursive: true });
}

const connection = new Connection('https://api.devnet.solana.com', 'confirmed');

async function createKeypair() {
    const keypair = Keypair.generate();
    const publicKey = keypair.publicKey.toBase58();
    const secretKey = JSON.stringify(Array.from(keypair.secretKey));

    fs.writeFileSync(KEYPAIR_PATH, secretKey);
    console.log(`Public Key: ${publicKey}`);
    console.log(`Private Key saved to ${KEYPAIR_PATH}`);
}

async function requestAirdrop(publicKeyString, amount) {
    const publicKey = new PublicKey(publicKeyString);
    const airdropSignature = await connection.requestAirdrop(publicKey, amount * LAMPORTS_PER_SOL);
    await connection.confirmTransaction(airdropSignature);
    console.log(`Airdropped ${amount} SOL to ${publicKeyString}`);
}

async function sendSol(recipientPublicKeyString, amount) {
    const secretKey = JSON.parse(fs.readFileSync(KEYPAIR_PATH));
    const senderKeypair = Keypair.fromSecretKey(Uint8Array.from(secretKey));
    const recipientPublicKey = new PublicKey(recipientPublicKeyString);

    const transaction = new Transaction().add(
        SystemProgram.transfer({
            fromPubkey: senderKeypair.publicKey,
            toPubkey: recipientPublicKey,
            lamports: amount * LAMPORTS_PER_SOL,
        })
    );

    const signature = await connection.sendTransaction(transaction, [senderKeypair]);
    await connection.confirmTransaction(signature);
    console.log(`Sent ${amount} SOL to ${recipientPublicKeyString}`);
}

async function main() {
    if (argv._.includes('create')) {
        await createKeypair();
    } else if (argv._.includes('airdrop')) {
        const publicKey = argv.publicKey;
        const amount = argv.amount;
        if (!publicKey || !amount) {
            console.error('Please provide a public key and amount for airdrop.');
            return;
        }
        await requestAirdrop(publicKey, amount);
    } else if (argv._.includes('send')) {
        const recipientPublicKey = argv.recipientPublicKey;
        const amount = argv.amount;
        if (!recipientPublicKey || !amount) {
            console.error('Please provide a recipient public key and amount to send.');
            return;
        }
        await sendSol(recipientPublicKey, amount);
    } else {
        console.log('Usage:');
        console.log('  create                  Create a new keypair');
        console.log('  airdrop --publicKey <publicKey> --amount <amount>   Request an airdrop');
        console.log('  send --recipientPublicKey <publicKey> --amount <amount> Send SOL to another public key');
    }
}

main().catch(err => {
    console.error(err);
});

File: solana_cli_tool.py
import argparse
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solana.rpc.api import Client
from solana.transaction import Transaction
from solders.system_program import TransferParams, transfer
import base58
import os

# Initialize the Solana Client
solana_client = Client("https://api.devnet.solana.com", "confirmed")

def create_keypair():
    keypair = Keypair()
    public_key = keypair.pubkey()
    private_key = keypair.secret()
    print(f"Public Key: {public_key}")
    print(f"Private Key: {base58.b58encode(private_key).decode()}")
    save_keypair(keypair)
    return keypair

def save_keypair(keypair):
    with open(os.path.expanduser('~/.config/solana/id.json'), 'w') as f:
        f.write(str(list(keypair.secret())))

def request_airdrop(public_key, amount):
    response = Client.request_airdrop(public_key, amount,  1000000000) # Convert SOL to lamports
    print(f"Airdrop response: {response}")

def send_sol(sender_keypair, recipient_public_key, amount):
    transaction = Transaction().add(
        transfer(
            TransferParams(
                from_pubkey=sender_keypair.public_key,
                to_pubkey=PublicKey(recipient_public_key),
                lamports=amount * 1000000000  # Convert SOL to lamports
            )
        )
    )
    response = solana_client.send_transaction(transaction, sender_keypair)
    print(f"Transaction response: {response}")

def main():
    parser = argparse.ArgumentParser(description="Solana CLI Tool")
    subparsers = parser.add_subparsers(dest="command")

    create_parser = subparsers.add_parser("create", help="Create a new keypair")

    airdrop_parser = subparsers.add_parser("airdrop", help="Request an airdrop")
    airdrop_parser.add_argument("public_key", type=str, help="Public key to airdrop SOL to")
    airdrop_parser.add_argument("amount", type=int, help="Amount of SOL to airdrop")

    send_parser = subparsers.add_parser("send", help="Send SOL to another public key")
    send_parser.add_argument("recipient_public_key", type=str, help="Recipient's public key")
    send_parser.add_argument("amount", type=int, help="Amount of SOL to send")

    args = parser.parse_args()

    if args.command == "create":
        create_keypair()
    elif args.command == "airdrop":
        request_airdrop(args.public_key, args.amount)
    elif args.command == "send":
        with open(os.path.expanduser('~/.config/solana/id.json'), 'r') as f:
            secret_key = eval(f.read())
        sender_keypair = Keypair.from_secret_key(bytes(secret_key))
        send_sol(sender_keypair, args.recipient_public_key, args.amount)
    else:
        parser.print_help()

if __name__ == "__main__":
    main()